### ISSUE TITLE
High CPU usage in empty project

### ISSUE THREAD

**Issue by randomPoison** [amethyst/amethyst#780]
When running an empty amethyst application (i.e. one with no systems registered and no extra logic happening in the game state's `update` method) I've noticed unexpectedly high CPU usage on my Windows 10, somewhere between 15%-25% with a 6 core CPU. I would expect to only see ~1% usage, since the application should be effectively waking up briefly at the beginning of each frame, and then immediately going back to sleep until the next frame. Note that this is limiting to 60 fps with a sleep-only limiting strategy, which should be the strategy the has the lowest CPU-usage.

I profiled the application, and found that most of the CPU time was being spent in rayon's worker thread logic. As best I can tell, what's happening is that rayon has a grace period where worker threads will spin-wait before sleeping them when there's no work available. Once more work is added, all sleeping worker threads are woken back up. Since amethyst adds new work every frame, the worker threads are being woken back up each frame, and once all the work for the frame is done, every worker thread has to spin-wait for some amount of time. This results in an (apparently) large amount of extra CPU time being eaten each frame.

I'm not able to test this behavior on MacOS/Linux, but I'd expect the result is the same on Unix-like platforms as it is on Windows.

It seems like this could be solved if amethyst could notify rayon that all the work for the thread is done, allowing rayon to immediately put all worker threads to sleep until the next frame. I don't know if this is something rayon supports, or if it's the optimal solution.

## How To Get Started

The biggest source of CPU usage when Amethyst is done with all work for a frame is coming from Rayon. The way Amethyst uses Rayon is triggering worst-case behavior in how Rayon handles putting worker threads to sleep, and the only real solution is for Rayon to add better support for quickly sleeping worker threads at the end of the frame.

A good first step for fixing this would be to build a demo that uses rayon the way that Amethyst does so that the Rayon devs can reproduce the issue and test fixes. Specifically, the demo should do the following:

1. Create a thread pool using Rayon. Ideally it should use a setup similar to what Amethyst does in terms of how many worker threads it creates.
2. Run in a loop that spawns a fixed amount of work (a good amount would be 5-10 milliseconds, bonus points for making the amount configurable) at the beginning of every loop iteration.
3. Each iteration, wait for work spawned on the worker thread to finish and then have the main thread wait until enough time has passed to start the next "frame". 60 fps is a good test case, so the demo should attempt to have the body of the loop take 16.6 ms each iteration.

Once we have this demo, we can make a PR to Rayon adding it to their demo projects, and then discuss with the Rayon team about how to best solve the problem.

**Comment by Rhuagh** [comment]
Ah, that makes perfect sense tbh. Sleeping is inherently a single thread construct, as is yield. And I'm not sure completely sleeping the thread pool is a good solution, that might affect things like asset loading, audio decompression etc.

**Comment by Rhuagh** [comment]
I can see no way to control the sleeping of the thread pool from the outside :/

**Comment by minecrawler** [comment]
I believe that having an empty Amethyst application is an edge-case. Usually, you'd have a lot of systems running, adding work to the pool. Putting the workers to sleep when all work is done and waking them up the next frame might put an overhead on the standard scenario, so I don't know how useful that kind of optimization actually is. I'd rather have workers waiting for work, putting a bit of stress on the CPU than having to go knock them out of bed for every little thing they should do ;)

**Comment by Rhuagh** [comment]
I tend to agree, in most scenarios that aren't bare bones examples you'll have enough work. I also think the grand majority of games will have vsync enabled which will kind of remove the need for a limiter because the game will be naturally limited.

**Comment by Xaeroxe** [comment]
We could alleviate both situations.  Right now we just allocate threads equal to your CPU core count.  We could do something like `cpu_core_count.min(actual_max_threads_dispatcher_will_ever_need + 1)` where the +1 is to help with other tasks such as asset loading.

**Comment by randomPoison** [comment]
So a couple of things I should clarify:

* I agree that trying to optimize an empty application isn't important, my worry is more that this case represents the minimum amount of CPU overhead added by frame limiting. For example, if my game is running at 60 fps but each frame only takes ~10 ms to complete, I'd expect to see ~60% CPU usage, but if I see closer to 90% then I'd be worried my game is wasting CPU resources unnecessarily.
* I also agree that for desktop and console gaming, it may well be worth leaving those worker threads active in order to maximize performance when there is work. But, for devices running on battery (i.e. mobile devices and laptops), wasted CPU time means wasted battery life, and it's often worth taking the hit in performance to avoid draining the battery.
* I don't think that the thread pool should be put to sleep if there are assets being loaded or other background work still happening. I'm only concerned with what happens when there is no more work being done. In those cases, all the worker threads sit there, idly spinning on the CPU with nothing to do.

That all being said, I agree that the fix for this is likely going to need some coordination from rayon, probably in the form of some kind of hinting that will allow rayon to sleep worker threads more efficiently in amethyst's use case.

**Comment by Rhuagh** [comment]
Not sure limiting the number of threads here will make much difference, the basic issue is that we can only sleep the main thread, not the threads in the thread pool.

**Comment by Xaeroxe** [comment]
I put something together real quick just now to see if my idea has any merit by getting the max thread use for various examples in Amethyst.

Pong: 5
Renderable: 5
Animation: 2
Arc_ball_camera: 3
fly_camera: 3
Material: 0 // Don't count this seriously, this example actually does nothing dynamic.
UI example: 6
Sprites: 5

In my personal box I have an 8 core CPU.  If we add 1 more thread to all of these the highest we get is 7 from the UI example.  A lot of our examples could be allocating fewer threads though.

**Comment by randomPoison** [comment]
@Xaeroxe Do you see that as still being beneficial if rayon were not using high CPU when idle? Or is reducing the number of threads only a stopgap fix for this particular issue?

**Comment by Xaeroxe** [comment]
My intent is to reduce the amount of idle threads so that this issue is less severe.  I'm not 100% certain that intentionally trying to put threads to sleep when work is done is in our best interest as it risks harming framerate when the threads need to be woken up again.

**Comment by Xaeroxe** [comment]
More idle thoughts: Let's say we have a game that uses 5 threads maximum.  At 5 threads it can achieve 1400 FPS (Not an exaggeration, we've seen Amethyst do this).  However the game is also V-Synced to 60 FPS.  That means about 5% of the frame execution time we're actually doing something.  95% of the time the processor is spinning needlessly.  That means there isn't really a good reason we shouldn't cut the thread pool down to two threads, one for gameplay and one for asset loading.  Then at this point we're only spinning 75% of the time on the gameplay thread, rather than 95% of the time on 5 gameplay threads.

**Comment by randomPoison** [comment]
My impression of the situation (though this needs confirmation, if possible) is that the worker threads *are* going to sleep each frame if there's enough idle time at the end of the frame, they just take too long to go to sleep. By the looks of the logic in rayon, it wouldn't take long for the threads to go to sleep, so I'd expect that they all end up asleep if you have more than a few milliseconds of idle time at the end of the frame. The issue isn't that the worker threads spin-wait nonstop during that idle time, rather that you have `num_cores` threads spin-waiting for a brief period at the end of each frame.

**Comment by Xaeroxe** [comment]
Ok I understand better now.  I still think we should try and reduce the thread count in Amethyst if we're not getting any benefit from having more threads, but I can see that's somewhat unrelated to the issue you're seeing.

**Comment by Xaeroxe** [comment]
Relevant reading: https://github.com/rayon-rs/rayon/blob/11bd211d3602719310ab6ed6c6e3ae5c876029ae/rayon-core/src/sleep/README.md

**Comment by Xaeroxe** [comment]
Sort of relevant: https://github.com/slide-rs/shred/pull/95

**Comment by termhn** [comment]
> I also think the grand majority of games will have vsync enabled which will kind of remove the need for a limiter because the game will be naturally limited.

I seem to be of the minority opinion within Rust gamedev circles, but I personally absolutely despise vsync except for in some very specific cases (for example a 2d fighting game with very specific frame animations or something similar). I also think any game that is not part of this very small group should have an adjustable frame rate limiter that goes to at least 0-300 fps so that gsync/freesync monitors up through 240hz can be supported properly. The problems with vsync are long and already enumerated elsewhere, but the most glaring to me are the often large and very significant additional input lag and the massive frame rate jumps (drops) if you drop even a frame or two below the monitor's refresh rate.

**Comment by Rhuagh** [comment]
Gsync/freesync is also a form of vsync, it's just a better algorithm for it.

**Comment by termhn** [comment]
Well, yes, I suppose, though at this point I think vsync is synonymous with the older algorithms to do so with "dumb" monitors especially within game settings menus so I don't generally include gsync/freesync in the same moniker :P

**Comment by Rhuagh** [comment]
It's stopping execution of the rendering and waiting for a signal, that's vsync. The fact that the wait time is adaptive instead of static is a pretty minor difference imho.

**Comment by torkleyy** [comment]
I don't think that creating less threads than cores makes sense. That might have a negative impact on performance, especially if one thread is blocked by some operation.

**Comment by Xaeroxe** [comment]
Yeah it could if not done carefully.  That's why I plan on asking the dispatcher first how many threads it can actually utilize.  Then if we add 2 to that number then we're basically guaranteed enough threads I think.  At that point adding more threads is just overhead door no good reason.  Not if we can't fully utilize them all anyways.

**Comment by OvermindDL1** [comment]
What server schedulers like the BEAM VM do is spool up a number of threads equal to the core count for the primary work in addition to a startup of 2 IO threads for IO waiting where it can spool up more on demand up to a set max (I think it defaults to a max of 20), although it rarely needs these as it uses async primitives in the great majority of cases.

**Comment by jojolepro** [comment]
I didn't read everything, but fixing this is important.
A game server running normally at 5 frames per second without a display to allow having many many instances at once wouldn't be possible with this bug, because each instance would take a whole cpu core just idling.

**Comment by Rhuagh** [comment]
Instances will only use CPU if NO other processes need CPU.

**Comment by randomPoison** [comment]
Well, that's kind of the concern here. It's good that amethyst will yield CPU time to other applications while idle, but it's also not ideal that amethyst is still using so much CPU time if no other applications are running.

**Comment by jojolepro** [comment]
Rip phone battery :/

**Comment by mdonoughe** [comment]
How is [this](https://github.com/mdonoughe/rayon/tree/game) for a demo? I added to the existing life demo a mode where it runs with a frame rate limiter and measures CPU usage instead of wall time. When I run it with the default options in release mode on my Ryzen 7 1800X with hyperthreading enabled (16 logical processors) in Windows 10:

      serial: 59.93 fps
        cpu usage: 4.7%
    parallel: 59.97 fps
      cpu usage: 55.3%
    par_bridge: 59.94 fps
      cpu usage: 1164.1%

By changing the size of the grid we change the amount of work done each frame. By changing the FPS we change the delay between frames. I'll collect some data about CPU usage for varying grid sizes and frame rates.

In WSL I get low values for the parallel stage CPU usage. I haven't tried running on a real Linux machine with a large number of processor threads to see if it's a WSL quirk.

**Comment by Moxinilian** [comment]
If you want to try it out in an Amethyst context, you can clone the repo, patch it with your rayon fork and run the `window` example in release mode (this is basically an empty example with a simple blue window) to see if there is any improvement.

Just in case you don't know that magic trick, you can patch the clone with your fork by adding this to the root Cargo.toml:

```toml
[patch.crates-io]
rayon = { git = "https://github.com/mdonoughe/rayon.git", branch = "game" }
```

**Comment by mdonoughe** [comment]
All I've done is set up a demo so unfortunately trying it in Amethyst won't make any difference.

Are there any notable differences between the way the demo uses Rayon and the way Amethyst uses Rayon? It seems to be exhibiting the same CPU usage issues when it's a small amount of work happening on a timer instead of a large amount of work to do as quickly as possible.

**Comment by randomPoison** [comment]
@mdonoughe That seems like a pretty good way of handling it, thanks for putting that together! Can you make a PR to Rayon and point them to this issue as the motivating case? That should get the discussion going with the Rayon devs and we can work out what the next steps are 🙂 

Also, it looks like you implemented some logic for measuring CPU usage. Have you tried the [cpu-monitor](https://crates.io/crates/cpu-monitor) crate? I haven't tried it myself, but it might save you some code.

**Comment by mdonoughe** [comment]
It looks like cpu-monitor is getting the system CPU time instead of the process CPU time.

I'll start writing up the pull request.

**Comment by OvermindDL1** [comment]
If you need a system with a decent number of physical cores I have a 16 native core (32 hyperthreads) at home that I can test with if you give me a set of commands (`cargo benchmark par_game_test` or whatever) to run.

**Comment by Rhuagh** [comment]
The problem is being discussed in https://github.com/rayon-rs/rayon/issues/576.

**Comment by jojolepro** [comment]
Moving to next release as this is an issue with a dependent crate and its not a critical issue.

**Comment by derekdreery** [comment]
@randomPoison Could you open an issue on [`cpu-monitor`](https://crates.io/crates/cpu-monitor) With a description of your use case (is it just cpu usage for a single process)? I like feature requests :D

**Comment by fhaynes** [comment]
This appears to have been resolved, closing.

**Comment by kabergstrom** [comment]
The Rayon issue is still present, but is mainly noticeable for CPUs with a large number of cores. The work-stealing algorithm used by `rayon` scales the work it does to steal based on the number of threads in the thread pool, and it repeats this work for many iterations before going to sleep. Hence the CPU usage in an empty project is a bigger issue the more cores you have. The environment variable `RAYON_NUM_THREADS` can be used to choose a smaller number of threads for now, until a more fitting threadpool is made for Amethyst.

The referenced PR that was merged in March made some good progress on certain platforms, but it did not change the default strategy from `Yield`. This means that the default behaviour is unchanged and is likely to still consume a lot of CPU time on some machines. For those that experience this, I recommend choosing `Sleep` as the frame limiter strategy.

**Comment by jojolepro** [comment]
So the issue is not fixed, only mitigated?

**Comment by kabergstrom** [comment]
I would say there's still potential for high CPU usage in empty projects, so yeah. I'd say there are two issues in play:
- The semantics of `[std::thread::yield_now](https://doc.rust-lang.org/std/thread/fn.yield_now.html)` do not guarantee that the CPU goes into a low power state. Same goes for `sleep(0)`: there is no guarantee that the CPU goes into a lower power state as it may just have the same behaviour as `yield_now`. Amethyst defaults to `Yield` as the default strategy, which can result in a full core being saturated for platforms where a low power state is not triggered by repeated `yield_now` calls.

- Amethyst's work queuing behaviour triggers a pathological case in Rayon thread pool's sleeping strategy, since Rayon assumes it should spin a little to check for work before sleeping. This could be made much more efficient by deeply integrating with sync primitives in `shred`'s dispatcher. In combination with the work-stealing approach which scales the time spent spinning with the number of threads in the pool can cause excessive CPU usage for users with many logical cores.

**Comment by Marwes** [comment]
Seeing this as well, on my 8-core (ryzen) CPU a simple example consumes 90% CPU time as reported by task manager :/. 

Tried changing the limiter to use sleep instead but it does not appear to help.

```
        world
            .res
            .fetch_mut::<amethyst::core::frame_limiter::FrameLimiter>()
            .set_rate(
                amethyst::core::frame_limiter::FrameRateLimitStrategy::Sleep,
                60,
            );

```

**Comment by Sydius** [comment]
Similarly, my 16-core CPU consumes 180% CPU time as reported by `top` even with a frame limit of 30 w/`FrameRateLimitStrategy::Sleep`. If I additionally set `RAYON_NUM_THREADS` to a low number, CPU usage drops significantly. The combination of the two on the tip-of-tree Amethyst works for me for now.

For fun, I plotted the CPU usage with different values of `RAYON_NUM_THREADS`: https://imgur.com/a/S1s3wTM

**Comment by fhaynes** [comment]
Re-opening this as it is still an issue, and putting it in for 0.12 milestone.

**Comment by varomix** [comment]
I guess this is the right place for this
this is just crazy, is using 80% of all my 32 threads, I only see that when I render Highend offline VFX renders, I'm just running the first part of the pong tutorial.
Is this something that can be avoided? 
if not I guess I will not continue to use this engine, is too hungry :)

Thanks
![image](https://user-images.githubusercontent.com/728059/56627737-53013080-6604-11e9-8c9d-c9b4067d1208.png)

**Comment by fhaynes** [comment]
@varomix This is definitely something we're going to fix; it will just take a bit of time, because it involves one of our dependencies. In the interim, have you tried setting the `RAYON_NUM_THREADS` environment variable?

**Comment by termhn** [comment]
@varomix just to make sure you know/to be clear--that CPU usage is not caused by Amethyst actually being slow and chugging super hard, it's an artifact of keeping threads awake in case "real" work is spawned... so making something more complex than the pong tutorial is not going to scale your CPU usage as though the pong tutorial was a benchmark, as it will just use more of the CPU time that is currently being used up while 'idle'

**Comment by varomix** [comment]
I set export RAYON_NUM_THREADS=1
same thing with one thread, can't understand why it has to use one thread 100% makes no sense to me.

**Comment by Moxinilian** [comment]
@varomix Just to make sure, are you running your game in release mode? Amethyst cannot be used in debug mode.

**Comment by randomPoison** [comment]
@varomix When you say

> I set export RAYON_NUM_THREADS=1
> same thing with one thread

Do you mean that you only see one core at 100% usage, or that you still see all of your cores at 100% usage?

---

> can't understand why it has to use one thread 100% makes no sense to me.

I'd like to provide a more detailed explanation about this, for @varomix as well as anyone else who's following this thread. What's happening here is a bit subtle and has to do with how the operating system schedules processes and threads to run on the cores of your CPU.

So, in the case where you've effectively disabled the thread pool with `RAYON_NUM_THREAD=1`, and you're running an empty game (i.e. no logic or anything, just a minimal setup with no logic), the main loop of the application effectively boils down to the following:

```rust
loop {
    thread::sleep(0);
}
```

Since you have no real game logic, pretty much the entirety of every frame of the game is spent repeatedly calling [`thread::sleep`](https://doc.rust-lang.org/std/thread/fn.sleep.html). Normally, when you call `sleep`, you specify an amount of time you want the thread to remain asleep, and the operating system will try to keep it asleep for at least that long.

The key thing for us is that "at least that long" part. In the case of an empty test project with no application logic, we don't really care if the thread sleeps for longer than we requested. However, under more normal running conditions, not having a hard guarantee on how long the thread will be asleep causes BIG problems with keeping a consistent cadence of frames for the game. If you have 2ms left at the end of your frame and you try to sleep for that full time, there's a good chance on some platforms that the thread will be asleep for several milliseconds too long, resulting in stuttering and framerate changes in your game.

To avoid this, we pass a duration of 0 to `sleep`. With a duration of 0, the thread immediately yields execution, allowing the operating system to give that CPU time to another thread/process. However, if there are no other threads/processes that needs the time, the game's thread resumes execution pretty much immediately.

The result is that one core of your CPU will always be in use, however **the game won't actually be blocking other applications from running on that core**. In practice, only having one CPU core spinning at 100% usage while the game is idle is the best case scenario we can hope for, since it's the minimum amount of CPU usage we can have while still keeping a consistent frame cadence.

Hopefully that helps clarify why you'll still see at least one core at 100% CPU (and also why that's not as big a problem as it might seem)!

**Comment by varomix** [comment]
thank you for the explanation 
yes, one core 100% usage

are you saying, when I add the logic this behaviour will be gone? or I will still have a thread using 100%?
thank you

**Comment by Sydius** [comment]
> In practice, only having one CPU core spinning at 100% usage while the game is idle is the best case scenario we can hope for, since it's the minimum amount of CPU usage we can have while still keeping a consistent frame cadence.

If that level of consistent frame cadence trumps power consumption. On mobile devices (including laptops), this can be unacceptable and it certainly is for my use case of a rather CPU-unintensive game. I didn't choose Rust and this engine so that I could do more with the CPU, but so that I could do less.

> are you saying, when I add the logic this behaviour will be gone? or I will still have a thread using 100%?

Try using the tip-of-tree version of Amethyst if you haven't already. I have the same CPU and the usage dropped to less than 1% after switching to tip-of-tree, setting `FrameRateLimitStrategy::Sleep` (I don't know if this part matters), and setting the environment variable to a low number.

**Comment by Marwes** [comment]
@randomPoison I am well aware that this doesn't (significantly) hinder other processes. That doesn't mean that the status quo is a good or even (in my opinion) acceptable solution. Seeing a game take up close to 100% of the CPU time regardless of how simple it is does not look good regardless of the reasons behind it (most people. (And without limiting rayons threads it will take up 100% of each cpu core as well).

Amethyst is the only game/game engine where I have seen this behavior so clearly it is possible to do better here. Maybe this issue isn't doesn't need to have the highest priority but as far as I see, a better solution will be need if amethyst is going to be production ready (where users not knowledgeable to how `sleep` etc works will see the CPU usage)

**Comment by torkleyy** [comment]
> I have the same CPU and the usage dropped to less than 1% after switching to tip-of-tree, setting `FrameRateLimitStrategy::Sleep`

The frame rate strategy definitely plays an important role here; also see [the `frame_limiter` API docs](http://www.amethyst.rs/doc/latest/doc/amethyst_core/frame_limiter/index.html#frame-rate-limiting-strategies).

> are you saying, when I add the logic this behaviour will be gone? or I will still have a thread using 100%?

The behaviour of sleep / yield differs on different platforms. That's why we made the frame rate limiter configurable. If `Yield` gives you 100% CPU usage, then adding logic won't decrease it, but rather more time will be spent executing actual game logic.

There unfortunately is no one perfect solution here; there are multiple with different trade-offs. What you can do now is select the best strategy for your use case. If you want very low CPU usage, you can also use `SleepAndYield`, which should give you the lowest power consumption, but can occasionally cause an increased delta time.

Now, what we can do is improve the default to make the power consumption lower. That's planned for the 0.12 release.

> Maybe  this issue isn't doesn't need to have the highest priority but as far as I see, a better solution will be need if amethyst is going to be  production ready (where users not knowledgeable to how sleep etc works will see the CPU usage)

I agree. Maybe a first improvement would be to make `Sleep` the default?

**Comment by Marwes** [comment]
> I agree. Maybe a first improvement would be to make Sleep the default?

Sleep or SleepAndYield doesn't seem help from my testing https://github.com/Marwes/gluon_specs/blob/24e330095e3c410baee0df0911b591a2bb191eb2/examples/asteroids/asteroids.rs#L92-L100

**Comment by torkleyy** [comment]
Okay, I see, thanks for the feedback. In that case your high CPU usage is caused not by our frame rate limiter, but rather by rayon's `ThreadPool`... There actually are two issues with the CPU usage here.

**Comment by randomPoison** [comment]
@varomix Have you tried switching the frame limiting behavior to `Sleep` or `SleepAndYield`? In combinations with `RAYON_NUM_THREADS=1`, that should give you very low CPU usage at the cost of having a less consistent frame cadence.

---

@Sydius @Marwes I completely agree that spinning on one CPU, even if yielding that time to other processes, isn't acceptable in all cases (especially mobile devices), sorry for implying otherwise :disappointed: We specifically provide [multiple frame limiting strategies](https://www.amethyst.rs/doc/latest/doc/amethyst_core/frame_limiter/index.html#frame-rate-limiting-strategies) to allow you to customize the behavior based on the needs of your game. As of now, all the options we offer skew towards emphasizing consistent frame times (i.e. even the `Sleep` approach still does `sleep(0)` as opposed to sleeping for a longer period of time, which would be beneficial for reducing CPU usage), so we could definitely stand to add more targeted options for minimizing CPU usage at the cost of frame timing.

Unfortunately, though, any changes to the frame limiting strategy right now is going to be overshadowed by the issues we're having with Rayon. Even if you use a better frame limiting strategy, Rayon will continue to spin on all other worker threads. As far as I know, we still don't have a concrete solution to this issue, but it looks like it's going to be prioritized for the 0.12 release.

---

> Amethyst is the only game/game engine where I have seen this behavior so clearly it is possible to do better here.

@Marwes do you have any specific examples you could provide of other game engines that handle this in a better way? I've looked into this a number of times, but there's no good, comprehensive explanation of how to better handle frame limiting that I've been able to find, so being able to reference other engines' implementation would be very helpful :grin:

**Comment by Marwes** [comment]
> do you have any specific examples you could provide of other game engines that handle this in a better way?

Afraid not, just extrapolating from what I can see in the task manager while playing other games.

**Comment by kabergstrom** [comment]
@Marwes I noticed that you are using 0.10 in the repo you linked:
https://github.com/Marwes/gluon_specs/blob/24e330095e3c410baee0df0911b591a2bb191eb2/Cargo.toml#L26

Does `Sleep` still not work as intended if you use `master` branch of Amethyst?

**Comment by Marwes** [comment]
> Does Sleep still not work as intended if you use master branch of Amethyst?

@kabergstrom Still 80-90% cpu usage.

**Comment by jamadazi** [comment]
@varomix The high cpu usage does not come from amethyst being cpu-hungry. It does not prevent you from having complex cpu-intensive game logic. You don't have to switch to another game engine out of fear that amethyst will not let you make your game because it ate all the cpu (it did not).

The high cpu usage comes from amethyst repeatedly checking if it is time for the next frame yet, when all the work for the current frame is finished. It does not affect the performance of your game (or other programs on your computer). The OS will show 100% cpu usage regardless of how much actual work/logic your game does.

If your game has very little logic, you will see 100% CPU, but very little of that will be your actual work, most will be spent by amethyst checking if it is time for the next frame yet. If your game has a lot of logic, you will also see 100% CPU, but now the CPU will be doing your actual work, instead of being busy with just doing simple checks over and over again until it is time for the next frame.

This high CPU usage does *not* affect performance.

That said, it needs to be fixed in the name of power saving. Until this is fixed, amethyst is very undesirable for mobile games, because the high cpu usage will eat all the battery.

**Comment by kabergstrom** [comment]
To be clear, many do not have this behavior of high CPU usage on their machines (depending on a number of factors it seems), and it's not clear whether the problem exists at all on mobile platforms. The current code does yield CPU time to the OS, and the behavior of not putting the CPU in a low power state is an OS-level decision.

**Comment by dseevr** [comment]
I tested a few of the examples on master, and I'm only seeing absurd CPU usage in debug builds.  Release builds work as expected.

**Comment by jojolepro** [comment]
Also riskable had the gltf example running on release using 3% of the cpu https://gfycat.com/cautiousbackchafer-game-development-amethyst-rust

**Comment by randomPoison** [comment]
Were there any changes made that would explain why this issue is now fixed? I'm a little worried about letting this ticket be closed if we haven't actually done anything to address the issue 😓

**Comment by jojolepro** [comment]
Test results using an 8 thread cpu running the gltf example on linux/x11/nvidia:
SleepAndYield 2ms, 60 fps: 50% cpu usage (100% = one thread)
SleepAndYield 2ms, 5 fps: 5% cpu usage

I will be testing on windows shortly.

**Comment by jojolepro** [comment]
Well this took a little while to get everything working on my old install...

Results on windows 7:
SleepAndYield 2ms, 60 fps: 0-5% total cpu usage
SleepAndYield 2ms, 5 fps: 0-5% total cpu usage

Conclusion: The bug where one cpu core would stay capped at 100% is now fixed. Tested on both linux and windows (osx should have the same result as linux normally).

**Comment by dseevr** [comment]
@jojolepro were those debug or release builds?

**Comment by jojolepro** [comment]
All release. Debug is expected to eat a bunch of cpu... but in my results it was still able to run at 60 fps at 50% on all threads.
